mars/services/scheduling/worker/execution.py

```diff
@@ -333,10 +333,8 @@ async def _run_subtask_once():
                     logger.debug('Wait for subtask to cancel timed out (%s). '
                                  'Start killing slot %d', subtask_info.kill_timeout, slot_id)
                     await slot_manager_ref.kill_slot(slot_id)
-                    # as slot is restarted, we does not need to trace it
                     sub_pool_address = await slot_manager_ref.get_slot_address(slot_id)
                     await mo.wait_actor_pool_recovered(sub_pool_address, self.address)
-                    slot_id = None
```

Comment:
Why is this removed here? Removing this will cause an unexpected `release_free_slot` call and may have unexpected consequences.

Reply:
Because the original logic has two roles to release the free slot, They may release one slot twice when fault occurred.
- The `BandSlotControlActor`
- The `SubtaskExecutionActor`

This PR makes the acquire and release logic in one role, the `SubtaskExecutionActor`. The `BandSlotControlActor` just register the slots, it does not actually release the slot.

The corner case is that, when the `SubtaskExecutionActor` get an error when executing subtask. e.g. the a `MarsError`, but the sub pool process crashes after the `MarsError`. The `SubtaskExecutionActor` don't know if the sub pool process really crashes. So, the slots may be released twice. Another fix is to kill the sub pool when any exceptions catched, but it does not fix the root cause: `two roles releasing slots`.

Reply:
Please refer to this test case: https://github.com/mars-project/mars/pull/2413/files#diff-b233da94b04ddd71d61754e6c850c2d25a97087d2f5cc62e6da40c13a31bfc16R241

Reply:
Then I would prefer adding `session_stid` to `release_free_slot` method and assert equality inside it. I'd prefer raising errors at user end to fail silently.

Reply:
Got it. I added some checks to `release_free_slot`.